src/backend/distributed/operations/repair_shards.c

```diff
@@ -287,7 +413,119 @@ citus_move_shard_placement(PG_FUNCTION_ARGS)
 
 
 /*
- * master_move_shard_placement is a wrapper function for old UDF name.
+ * EnsureEnoughDiskSpaceForShardMove checks that there is enough space for
+ * shard moves of the given colocated shard list from source node to target node.
+ * It tries to clean up old shard placements to ensure there is enough space.
+ */
+static void
+EnsureEnoughDiskSpaceForShardMove(List *colocatedShardList,
+								  char *sourceNodeName, uint32 sourceNodePort,
+								  char *targetNodeName, uint32 targetNodePort)
+{
+	int64 colocationSize = ColocationSize(colocatedShardList, sourceNodeName,
+										  sourceNodePort);
+
+	uint32 connectionFlag = 0;
+	MultiConnection *connection = GetNodeConnection(connectionFlag, targetNodeName,
+													targetNodePort);
+	int checkSpaceTries = 0;
+	while (true)
+	{
+		bool droppedAllShards = false;
+		if (checkSpaceTries > 0)
+		{
+			/*
+			 * Only run shard cleanup when the first round of checks has failed
+			 * otherwise simply let the automatic cleanup do its job.
+			 *
+			 * NOTE: this tries to drop all marked shards, not only on the shard
+			 * that we're trying to move to.
+			 */
+			ereport(NOTICE, (errmsg("trying to drop shards to resolve issue...")));
+
+			bool waitForCleanupLock = true;
+			droppedAllShards = DropMarkedShards(waitForCleanupLock, NULL);
+
+			/*
+			 * NOTE: It actually doesn't make sense to wait in some cases where
+			 * it does droppedAllShards is false. Most importantly the case
+			 * where there are no shards marked for removal on the target
+			 * shard. It's not a big issue though, since in this case we simply
+			 * error out after a couple of more rounds of trying to drop the
+			 * left over shards.
+			 */
+		}
+
+		DeferredErrorMessage *temporaryError = CheckSpaceConstraints(connection,
+																	 colocationSize);
+		if (temporaryError == NULL)
+		{
+			break;
+		}
+
+		if (droppedAllShards)
+		{
+			/*
+			 * If we dropped all shards but still don't have enough space for shard move
+			 * then we don't have any choice but error
+			 */
+			RaiseDeferredError(temporaryError, ERROR);
+		}
+
+		RaiseDeferredError(temporaryError, WARNING);
+		checkSpaceTries++;
+		if (checkSpaceTries >= WaitForDeferShardsMaxTries)
+		{
+			ereport(ERROR, (
+						errmsg(
+							"shards were not removed after trying %d times, giving up",
+							checkSpaceTries),
+						errhint("Try increasing citus.move_shard_defer_drop_tries")
+						));
+		}
+
+		/* wait the timeout and try again*/
+		pg_usleep(WaitForDeferShardRetryTimeInSec * 1000 * 1000);
```

Comment:
sleeping might not be the most optimal way because with it you cannot read the signals that happen before the loop finishes. See `WaitForMiliseconds ()` in the enterprise repo

Reply:
WaitForMilliseconds makes sense, I see some other usages of pg_usleep in the codebase. Is there a downside of using `WaitForMilliseconds` compared to `pg_usleep`

Reply:
As far as I know, no there is not a downside of using it. Still, would be good to double check the parameters that we pass to `WaitLatch` in `WaitForMilliseconds`